src/net.cpp

```diff
@@ -651,26 +698,37 @@ bool CNode::ReceiveMsgBytes(const char *pch, unsigned int nBytes)
             {
                 nActivityBytes += msg.hdr.nMessageSize;
 
-                // If the message is a priority message then move from the back to the front of the deque.
-                //
-                // NOTE: for GET_XTHIN we don't jump the queue on test environments because the GET_XTHIN can get ahead
-                // of a previous GET_XTHIN/HEADER requests and result in a DOS if the block returns out of order and
-                // with no headers in the block index or the setblockindexcandidates.
-                if ((strCommand == NetMsgType::GET_XTHIN && Params().NetworkIDString() == "main") ||
-                    strCommand == NetMsgType::GET_THIN || strCommand == NetMsgType::XTHINBLOCK ||
-                    strCommand == NetMsgType::THINBLOCK || strCommand == NetMsgType::XBLOCKTX ||
-                    strCommand == NetMsgType::GET_XBLOCKTX || strCommand == NetMsgType::GET_GRAPHENE ||
-                    strCommand == NetMsgType::GRAPHENEBLOCK || strCommand == NetMsgType::GRAPHENETX ||
-                    strCommand == NetMsgType::GET_GRAPHENETX)
+                // If the message is a priority message then move it into the priority queue.
+                if (IsPriorityMsg(strCommand))
                 {
-                    LOG(THIN | GRAPHENE, "ReceiveMsgBytes %s\n", strCommand);
-
-                    // Move the this message to the front of the queue.
-                    std::rotate(vRecvMsg.begin(), vRecvMsg.end() - 1, vRecvMsg.end());
-
-                    std::string strFirstMsgCommand = vRecvMsg[0].hdr.GetCommand();
-                    DbgAssert(strFirstMsgCommand == strCommand, );
-                    LOG(THIN | GRAPHENE, "Receive Queue: pushed %s to the front of the queue\n", strFirstMsgCommand);
+                    LOCK(cs_PriorityRecvQ);
+                    // As a safeguard on some peer trying to dominate our networking
+                    // don't allow unlimited simultaneous priority messages. There should in general only be
+                    // one priority message at any one time with the exception of block HEADERS announcements
+                    // which can come relatively close together in time.
+                    if (vPriorityRecvQ.size() <= 5)
+                    {
+                        // Move the this message to the priority queue.
+                        CNetMessage priority;
+                        std::swap(priority, msg);
+                        vRecvMsg.pop_back();
```

Comment:
It was always a little weird to have an incomplete message hanging off the end of the vRecvMsg queue since that message cannot be processed.  Now that we place the message on one of two queues upon completion its even weirder because we need to pull it off of vRecvMsg if its a priority message.  I think that we should clean this up.

It would be just a few LOC to have a member variable named something like "msgInProgress" that holds the incomplete message, and we would add it to vPriorityRecvQ and assign a new empty one (in a manner similar to line 619) only when its complete as part of this if statement, or to vRecvMsg in an "else" path that's taken if either of these nested ifs fails.

Reply:
Not sure I understand the problem your outlining... The msg is complete when we swap and move it to the priority queue...it has no extra data dangling off the end,  msg.readData ensures we end up with only the full message and nothing more before going to through the next iteration of reading in the next message.